src/storage/src/source/generator/ldbc.rs

```diff
@@ -0,0 +1,247 @@
+                        },
+                    ));
+                    // Make symmetric.
+                    if csv.output == LDBC_PERSON_KNOWS_PERSON_OUTPUT {
+                        assert!(self.pending.is_none());
+                        let mut iter = row.iter();
+                        let creationdate = iter.next().expect("must exist");
+                        let person1id = iter.next().expect("must exist");
+                        let person2id = iter.next().expect("must exist");
+                        let swapped = Row::pack_slice(&[
+                            creationdate,
+                            person2id,
+                            person1id,
+                        ]);
+                        self.pending = Some((csv.output, swapped));
+                    }
+                    Some((csv.output, row))
+                }
+                Ok(false) => None,
+                Err(_) => None,
+            })
+    }
+}
+
+impl Iterator for Context {
+    type Item = (usize, Event<Option<MzOffset>, (Row, i64)>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((output, row)) = self.pending.take() {
+            self.offset += 1;
+            return Some((output, Event::Message(self.offset, (row, 1))));
+        }
+        loop {
+            match self.next_record() {
+                Some((output, row)) => {
+                    self.offset += 1;
+                    return Some((output, Event::Message(self.offset, (row, 1))));
+                }
+                None => {
+                    // If we're out of records for this file, close the current file and its data.
+                    if let Some(csv) = self.rdr.take() {
+                        self.progress_outputs.push(csv.output);
+                    }
+                    // Otherwise open the next file.
+                    self.next_file();
+                    // No more files.
+                    if self.rdr.is_none() {
+                        if let Some(output) = self.progress_outputs.pop() {
+                            return Some((output, Event::Progress(None)));
+                        }
+                        return None;
+                    }
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+/// Fetches URLs whose bodies are zstd compressed and returns their response body uncompressed and
+/// concatenated.
+async fn fetch_urls_compressed(urls: &[String]) -> Result<Vec<u8>, anyhow::Error> {
+    let mut buf = Vec::new();
+    for url in urls {
+        let resp = reqwest::get(url).await?.error_for_status()?;
+        let stream = resp.bytes_stream();
+        let stream = stream.map(|result| {
+            result.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+        });
+        let reader = StreamReader::new(stream);
+        let mut decoder = ZstdDecoder::new(reader);
+        tokio::io::copy(&mut decoder, &mut buf).await?;
```

Comment:
This will buffer the full uncompressed data in memory. Can we use a streaming decoder for both the zstd compression and the csv decoding?